### PR TITLE
Improve initial page load

### LINKED ISSUE #1357: Pages are loading twice
**Describe the bug**
When you open some DSpace 7 pages, they appear to load twice, which is confusing for the user.

I have been able to reproduce this at least on homepage / collection page / item page even if it doesn't happen systematically.
This behaviour can be experienced on demo7 : https://demo7.dspace.org/home

**To Reproduce**
1. Open https://demo7.dspace.org/home (or other collection / item pages) in a new browser tab
2. The page is first displayed for a short time
3. Then a blank page appears for a short time too
4. Finally the page reappears and is displayed properly

**Expected behavior**
Pages should only be loading once

**Related work**
@artlowel has already identified that 
- the first page displayed is the server side rendered version
- then JS kicks in and it starts loading
- then you see the client side rendered version

### LINKED ISSUE #1650: "Import" menu sometimes disappears from sidebar
**Describe the bug**
After login as an Admin, the "Import" menu initially appears.  However, if you select any other page in the sidebar, it disappears.  Here's a screenshot... normally Import is right above Export.

no-import-menu

Reproducible both locally & on http://demo7.dspace.org

**To Reproduce**
1. Login as an Admin
2. Look at sidebar. The Import menu should be there
3. Click on a *different* menu in sidebar, e.g. "Admin Search"
4. Look again at the sidebar, the Import menu will have disappeared.
5. If you click the reload button in your browser the Import menu will return.  It will disappear again after you click a different item in the menu.

It almost seems like some sort of caching issue in the UI layer?  Uncertain what could be causing this and why it's specific to the Import menu (no other menus are affected).

### PR DESCRIPTION
**Pull request by ybnd**
## References
* Fixes #1357
* Fixes #1650

## Description

Various improvements to make initial page loading faster & more smooth

* Compression etc.
  * Pre-compress static files instead of compressing all responses on the fly (keep compressing SSR responses though)
  * The default banner image is rather large → serve WEBP alongside JPG & provide smaller versions
* Ensure the loading animation is not shown on first load
* Restructure theme modules a bit
  * `EagerThemeModule`s declare components that are shown on every page & entry components
  * `LazyThemeModule`s contain all the rest
* Make sure the SSR/CSR swap only happens _after_ routing is complete via `initialNavigation: 'enabledBlocking'`
* Add i18n files to the transfer state so they won't have to be downloaded again on CSR
* Move public & admin menus to `MenuResolver` so they're created before routing completes

## Instructions for Reviewers
Start in production mode
  * The "double loading" issue should be resolved
  * Initial loading should be a bit quicker and less janky in general

## Checklist

_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
  * About half of the changes are due to the admin menu that was moved to another file, but remained mostly the same
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

**Comment by ybnd** [comment]
@tdonohue thanks for the tip, I'll check it out!
I narrowed it down yesterday to the `initialNavigation: 'enabledBlocking'` setting we introduced [here](https://github.com/DSpace/dspace-angular/blob/34be1179ab29ac264ecd8f0585de152dc4b155b9/src/app/app-routing.module.ts#L226): 
seems like Cypress doesn't wait long enough for pages to stabilize anymore. I've seen some cases where clicks actually connect (e.g. to the user menu dropdown) but don't have any effect, as if the page  isn't _fully_ loaded yet.

File: .gitignore [hunk 1]
```diff
@@ -37,3 +37,5 @@ package-lock.json
 
 .env
 /nbproject/
+
+junit.xml
```
Comment (tdonohue) [inline .gitignore:41]: Was this a purposeful change?  Or maybe accidental? Just seems unrelated to the rest of this PR.

File: src/app/shared/utils/browser-only.pipe.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import { isPlatformBrowser } from '@angular/common';
+
+/**
+ * A pipe that only returns its intput when run in the browser.
+ * Used to distinguish client-side rendered content from server-side rendered content.
+ */
+@Pipe({
+  name: 'dsBrowserOnly'
+})
+export class BrowserOnlyPipe implements PipeTransform {
```
Comment (tdonohue) [inline src/app/shared/utils/browser-only.pipe.ts:11]: First off, glad you figured out a way to work around the e2e test issues!

Could we provide a more detailed description of this new `dsBrowserOnly` pipe?  Correct me if I'm wrong, but it appears this might now be needed for **every component/button** you want to interact with in e2e tests?  Or is it somehow only needed for some e2e tests?

Based on your recent changes, it appears to me that this might be needed anywhere we add a `data-test` attribute, like:
```
[attr.data-test]="'value' | dsBrowserOnly"
```
But, I'm not sure if I'm overlooking something.  I also saw in the code commit that you mentioned this is used to essentially "block" e2e tests from running until Client side rendering (CSR) kicks in.  

Overall, I think a detailed description is warranted. You've got a good start of a description here, but it could be improved with details of how to use this for e2e tests & why it's necessary.

File: src/themes/custom/lazy-theme.module.ts [hunk 3]
```diff
@@ -66,32 +70,22 @@ import { ForgotEmailComponent } from './app/forgot-password/forgot-password-emai
...
 import { BreadcrumbsComponent } from './app/breadcrumbs/breadcrumbs.component';
-import { HeaderNavbarWrapperComponent } from './app/header-nav-wrapper/header-navbar-wrapper.component';
-import { FileSectionComponent } from './app/item-page/simple/field-components/file-section/file-section.component';
-import { SearchModule } from '../../app/shared/search/search.module';
-import { ResourcePoliciesModule } from '../../app/shared/resource-policies/resource-policies.module';
-import { ComcolModule } from '../../app/shared/comcol/comcol.module';
 import { FeedbackComponent } from './app/info/feedback/feedback.component';
 import { CommunityListComponent } from './app/community-list-page/community-list/community-list.component';
 
+// todo: optimize imports
```
Comment (tdonohue) [inline src/themes/custom/lazy-theme.module.ts:83]: Should we create a separate ticket for this TODO task? Or is this sufficiently resolved? (Not sure what the TODO is reminding us to actually do)

**Review by tdonohue — APPROVED** [review]
@ybnd : This looks great overall!  I've tested this and it works well... the "flash" of the page reloading is gone _almost_ everywhere.

That said, I still do see the old "flash" / reload behavior when switching the language in the header of the UI (at least the entire header flashes...the page sometimes doesn't).  However, that's the ONLY scenario where it's obviously visible (at least when I tested via either Chrome or Edge browsers).  So, even if it only appears in that one area, this PR provides much better behavior. 

👍 Overall, I approve of this work & it works well.  But, I do have a few (very minor) suggestions/questions inline below.

File: src/themes/custom/lazy-theme.module.ts (hunk 3 above)
Comment (ybnd) [reply src/themes/custom/lazy-theme.module.ts:83]: I think that's probably a note to myself I added when merging `main` & forgot to remove

File: src/app/shared/utils/browser-only.pipe.ts (hunk 2 above)
Comment (ybnd) [reply src/app/shared/utils/browser-only.pipe.ts:11]: On it!

> Correct me if I'm wrong, but it appears this might now be needed for every component/button you want to interact with in e2e tests? Or is it somehow only needed for some e2e tests?

I don't think we need `dsBrowserOnly` on _every_ button/input;
Only about 5 suites were failing in the first place, e.g. `submission.spec.ts` contains a bunch of interactions that were completely fine apparently, to be honest I don't know what makes the difference.

> I also saw in the code commit that you mentioned this is used to essentially "block" e2e tests from running until Client side rendering (CSR) kicks in. Is that accurate?

Yes.
The pipe ensures that SSR HTML won't contain the attribute, so Cypress will wait until CSR is ready for that particular component before interacting with it.

File: .gitignore (hunk 1 above)
Comment (artlowel) [reply .gitignore:41]: I added that, junit.xml is generated for an internal test coverage check at Atmire, and I saw a couple of instances of devs accidentally committing it in contribution branches and then having to remove it. It can't hurt to add it to gitignore imo.

File: src/app/shared/utils/browser-only.pipe.ts (hunk 2 above)
Comment (artlowel) [reply src/app/shared/utils/browser-only.pipe.ts:11]: >The pipe ensures that SSR HTML won't contain the attribute, so Cypress will wait until CSR is ready for that particular component before interacting with it.

To elaborate on that: those particular tests interact with components that need JS to function, so they wouldn't work on the pure SSR HTML. That's why we needed a way to differentiate between the SSR and the CSR version to know when to perform the test

File: src/app/shared/utils/browser-only.pipe.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * When rendered client-side, the HTML will contain `data-test="something"`
+ *
+ * This can be useful for end-to-end testing elements that need JS (that isn't included in SSR HTML) to function:
+ * By depending on `dsBrowserOnly` attributes in tests we can make sure we wait
+ * until such components are fully interactive before trying to interact with them.
+ */
+@Pipe({
+  name: 'dsBrowserOnly'
+})
+export class BrowserOnlyPipe implements PipeTransform {
```
Comment (atarix83) [inline src/app/shared/utils/browser-only.pipe.ts:22]: For what I understand here this pipe has been added to solve issues with e2e and SSR. If so we should mind that the same approach will be used in the future. So maybe we should mention it in the `Writing E2E Tests` section of the README file

File: src/themes/custom/eager-theme.module.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * This module is included in the main bundle that gets downloaded at first page load. So it should
+ * contain only the themed components that have to be available immediately for the first page load,
+ * and the minimal set of imports required to make them work. Anything you can cut from it will make
+ * the initial page load faster, but may cause the page to flicker as components that were already
+ * rendered server side need to be lazy-loaded again client side
+ *
+ * Themed EntryComponents should also be added here
+ */
+export class EagerThemeModule {
```
Comment (atarix83) [inline src/themes/custom/eager-theme.module.ts:57]: I think this change should be well documented also in the official documentation after this PR will be merged https://wiki.lyrasis.org/display/DSDOC7x/User+Interface+Customization#UserInterfaceCustomization-CustomizeotherComponentsinyourTheme

Could you create an issue to keep track of it?

**Review by atarix83 — CHANGES_REQUESTED** [review]
thanks @ybnd for this PR

I've test it, it's overall good and it seems a very notable improvement. I've only just a few inline feedback added

**Comment by tdonohue** [comment]
@ybnd : just letting you know this has a small merge conflict now that needs resolution.

**Comment by tdonohue** [comment]
@atarix83 : Could you give this another review?  It looks like your feedback has been addressed, except for adding a note on Eager/Lazy Themed components to our documentation (but my assumption is we can ask @ybnd or @artlowel to add that note once this is merged).

**Comment by ybnd** [comment]
@tdonohue I've written up some bits for that but I don't have a wiki account myself (tried requesting one a while back but didn't get a reply)

Stuff to change:

* `theme.module.ts` → `lazy-theme.module.ts`*

* *[Theme directories](https://wiki.lyrasis.org/display/DSDOC7x/User+Interface+Customization#UserInterfaceCustomization-ThemeDirectories&DesignPrinciples) (or maybe [Design principles](https://wiki.lyrasis.org/display/DSPACE/DSpace+UI+Design+principles+and+guidelines)?) → I'd add a part about the "anatomy" of a theme, something like*

  A theme's directory should include the following files and directories

  * `app/` contains the theme's Angular components and should mirror the structure of `src/app/`
  * `assets/` contains the theme's custom assets, such as fonts or images
  * `styles/` contains the theme's global styles
  * `eager-theme.module.ts` declares the components that should be included in the app's main bundle, such as
    * *Eager components* are those that should be available immediately when first loading, such as the main parts of the homepage and components that are present on every page.
    * [*Entry components*](https://angular.io/guide/entry-components) that are registered via a decorator such as `@listableObjectComponent`
  * `lazy-theme.module.ts` declares all the components of the theme. 

* *[Getting started](https://wiki.lyrasis.org/display/DSDOC7x/User+Interface+Customization#UserInterfaceCustomization-GettingStarted) → new step (somewhere after 3):*

  Import the new theme's `eager-theme.module.ts` in `themes/eager-themes.module.ts`. If you're switching from one theme to another, remove the old theme from the imports.

* *[Adding components to a theme](https://wiki.lyrasis.org/display/DSDOC7x/User+Interface+Customization#UserInterfaceCustomization-AddingComponentDirectoriestoyourTheme) → new step after 3:*

  Some components should also be included in the theme's `eager-theme.module.ts`

  * Eager components should be added to `DECLARATIONS`
  * Entry components should be added to `ENTRY_COMPONENTS`

* *[Removing components from a theme](https://wiki.lyrasis.org/display/DSDOC7x/User+Interface+Customization#UserInterfaceCustomization-RemovingComponentDirectoriesfromyourTheme)*

  > remove all references to that directory/component from your theme's `theme.module.ts` ~~file~~

  ... and `eager-theme.module.ts` files.

File: src/themes/custom/eager-theme.module.ts (hunk 5 above)
Comment (ybnd) [reply src/themes/custom/eager-theme.module.ts:57]: Updated the wiki

**Review by atarix83 — APPROVED** [review]
thanks @ybnd  

LGTM

**Comment by tdonohue** [comment]
Merging as this is at +2 now. Thanks again @ybnd !
